cmd/finalize/finalize.go

```diff
@@ -79,61 +77,124 @@ func main() {
+				// Nack errorus message so the server gets notified that something is wrong but don't requeue the message
+				if e := delivered.Nack(false, false); e != nil {
+					log.Errorln("failed to Nack message, reason: ", e)
+				}
+				// Send the errorus message to an error queue so it can be analyzed.
+				josnError := broker.JSONError{
+					Error:          "Validation of json message failed",
+					Reason:         fmt.Sprintf("%v", res.Errors()),
+					OrginalMessage: delivered.Body,
+				}
+				body, _ := json.Marshal(josnError)
 
-			if err := json.Unmarshal(delivered.Body, &message); err != nil {
-				log.Errorf("Unmarshaling json message failed, reason: %s", err)
-				// publish MQ error
+				if e := broker.SendMessage(mq, delivered.CorrelationId, conf.Broker.Exchange, conf.Broker.RoutingError, conf.Broker.Durable, body); e != nil {
+					log.Error("faild to publish message, reason: ", e)
+				}
+				// Restart on new message
 				continue
 			}
 
+			// we unmarshal the message in the validation step so this is safe to do
+			_ = json.Unmarshal(delivered.Body, &message)
+
 			// Extract the sha256 from the message and use it for the database
 			var checksumSha256 string
 			for _, checksum := range message.DecryptedChecksums {
 				if checksum.Type == "sha256" {
 					checksumSha256 = checksum.Value
 				}
 			}
-			log.Debug("Mark ready")
-			if err := db.MarkReady(message.AccessionID, message.User, message.Filepath, checksumSha256); err != nil {
-				log.Errorf("MarkReady failed, reason: %v", err)
-				continue
-				// this should be handled by the SQL retry mechanism
-			}
-			c := Completed{
+
+			c := completed{
 				User:               message.User,
 				Filepath:           message.Filepath,
 				AccessionID:        message.AccessionID,
 				DecryptedChecksums: message.DecryptedChecksums,
 			}
 
-			completeMsg := gojsonschema.NewReferenceLoader(conf.SchemasPath + "ingestion-completion.json")
-			res, err = gojsonschema.Validate(completeMsg, gojsonschema.NewGoLoader(c))
+			completeMsg, _ := json.Marshal(&c)
+
+			res, err = validateJSON(conf.SchemasPath, completeMsg)
 			if err != nil {
-				fmt.Println("error:", err)
-				log.Error(err)
-				// publish MQ error
+				log.Errorf("josn error: %v", err)
+				// Nack errorus message so the server gets notified that something is wrong but don't requeue the message
+				if e := delivered.Nack(false, false); e != nil {
+					log.Errorln("failed to Nack message, reason: ", e)
+				}
+				// Send the errorus message to an error queue so it can be analyzed.
+				josnError := broker.JSONError{
+					Error:          "Handling json message failed",
+					Reason:         err.Error(),
+					OrginalMessage: delivered.Body,
+				}
+				body, _ := json.Marshal(josnError)
+
+				if e := broker.SendMessage(mq, delivered.CorrelationId, conf.Broker.Exchange, conf.Broker.RoutingError, conf.Broker.Durable, body); e != nil {
+					log.Error("faild to publish message, reason: ", e)
+				}
```

Comment:
Can't this be made a function in broker rather than having this much code for messages everywhere)?

Reply:
I was thinking about it